File: apps/ios_native/Sources/PasswordManageriOSCore/Models/VaultModels.swift
import Foundation

enum VaultEntryType: String, CaseIterable, Codable, Identifiable, Sendable {
    case credential
    case server
    case service

    var id: String { rawValue }

    var title: String {
        switch self {
        case .credential: "Credential"
        case .server: "Server"
        case .service: "Service"
        }
    }

    var systemImage: String {
        switch self {
        case .credential: "key"
        case .server: "server.rack"
        case .service: "network"
        }
    }
}

struct CredentialPayload: Codable, Equatable, Sendable {
    var username: String = ""
    var password: String = ""
    var token: String = ""
    var appId: String = ""
    var accessKey: String = ""
    var secretKey: String = ""
    var notes: String = ""
    var tags: [String] = []
    var category: String = ""
}

struct ServerPayload: Codable, Equatable, Sendable {
    var name: String = ""
    var ipAddress: String = ""
    var port: String = ""
    var username: String = ""
    var password: String = ""
    var basicConfig: String = ""
    var operatingSystem: String = ""
    var location: String = ""
    var notes: String = ""
    var tags: [String] = []
    var accountId: String?
    var category: String = ""
}

struct ServiceAccount: Codable, Equatable, Identifiable, Sendable {
    var id = UUID()
    var username: String = ""
    var password: String = ""
    var note: String = ""

    private enum CodingKeys: String, CodingKey {
        case username
        case password
        case note
    }

    init(
        id: UUID = UUID(),
        username: String = "",
        password: String = "",
        note: String = ""
    ) {
        self.id = id
        self.username = username
        self.password = password
        self.note = note
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = UUID()
        username = try container.decodeIfPresent(String.self, forKey: .username) ?? ""
        password = try container.decodeIfPresent(String.self, forKey: .password) ?? ""
        note = try container.decodeIfPresent(String.self, forKey: .note) ?? ""
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(username, forKey: .username)
        try container.encode(password, forKey: .password)
        try container.encode(note, forKey: .note)
    }
}

struct ServicePayload: Codable, Equatable, Sendable {
    var name: String = ""
    var connectionAddress: String = ""
    var connectionPort: String = ""
    var accountId: String?
    var serverIds: [String] = []
    var accounts: [ServiceAccount] = []
    var notes: String = ""
    var tags: [String] = []
    var category: String = ""
}

struct CustomField: Identifiable, Codable, Equatable, Sendable {
    var id: UUID = UUID()
    var name: String = ""
    var value: String = ""

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case value
    }

    init(
        id: UUID = UUID(),
        name: String = "",
        value: String = ""
    ) {
        self.id = id
        self.name = name
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        name = try container.decodeIfPresent(String.self, forKey: .name) ?? ""
        value = try container.decodeIfPresent(String.self, forKey: .value) ?? ""
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id.uuidString.lowercased(), forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(value, forKey: .value)
    }
}

enum VaultPayload: Codable, Equatable, Sendable {
    case credential(CredentialPayload)
    case server(ServerPayload)
    case service(ServicePayload)

    private enum CodingKeys: String, CodingKey {
        case credential
        case server
        case service
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let payload = try container.decodeIfPresent(CredentialPayload.self, forKey: .credential) {
            self = .credential(payload)
            return
        }
        if let payload = try container.decodeIfPresent(ServerPayload.self, forKey: .server) {
            self = .server(payload)
            return
        }
        if let payload = try container.decodeIfPresent(ServicePayload.self, forKey: .service) {
            self = .service(payload)
            return
        }
        throw DecodingError.dataCorrupted(
            DecodingError.Context(
                codingPath: decoder.codingPath,
                debugDescription: "Expected credential, server, or service payload."
            )
        )
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .credential(let payload):
            try container.encode(payload, forKey: .credential)
        case .server(let payload):
            try container.encode(payload, forKey: .server)
        case .service(let payload):
            try container.encode(payload, forKey: .service)
        }
    }

    var category: String {
        switch self {
        case .credential(let payload): payload.category
        case .server(let payload): payload.category
        case .service(let payload): payload.category
        }
    }

    var tags: [String] {
        switch self {
        case .credential(let payload): payload.tags
        case .server(let payload): payload.tags
        case .service(let payload): payload.tags
        }
    }
}

struct VaultEntry: Identifiable, Codable, Equatable, Sendable {
    var id: UUID = UUID()
    var label: String
    var type: VaultEntryType
    var payload: VaultPayload
    var customFields: [CustomField] = []
    var createdAt: Date = Date()
    var updatedAt: Date = Date()
    var isDeleted: Bool = false
    var deletedAt: Date?
    var version: [String: Int] = [:]
    var updatedBy: String = "macos-native"

    private enum CodingKeys: String, CodingKey {
        case id
        case label
        case type
        case payload
        case customFields
        case createdAt
        case updatedAt
        case isDeleted
        case deletedAt
        case version
        case updatedBy
    }

    init(
        id: UUID = UUID(),
        label: String,
        type: VaultEntryType,
        payload: VaultPayload,
        customFields: [CustomField] = [],
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        isDeleted: Bool = false,
        deletedAt: Date? = nil,
        version: [String: Int] = [:],
        updatedBy: String = "ios-native"
    ) {
        self.id = id
        self.label = label
        self.type = type
        self.payload = payload
        self.customFields = customFields
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.isDeleted = isDeleted
        self.deletedAt = deletedAt
        self.version = version
        self.updatedBy = updatedBy
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        label = try container.decode(String.self, forKey: .label)
        type = try container.decode(VaultEntryType.self, forKey: .type)
        payload = try container.decode(VaultPayload.self, forKey: .payload)
        customFields = try container.decodeIfPresent([CustomField].self, forKey: .customFields) ?? []
        createdAt = try container.decodeIfPresent(Date.self, forKey: .createdAt) ?? Date()
        updatedAt = try container.decodeIfPresent(Date.self, forKey: .updatedAt) ?? Date()
        isDeleted = try container.decodeIfPresent(Bool.self, forKey: .isDeleted) ?? false
        deletedAt = try container.decodeIfPresent(Date.self, forKey: .deletedAt)
        version = try container.decodeIfPresent([String: Int].self, forKey: .version) ?? [:]
        updatedBy = try container.decodeIfPresent(String.self, forKey: .updatedBy) ?? "ios-native"
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id.uuidString.lowercased(), forKey: .id)
        try container.encode(label, forKey: .label)
        try container.encode(type, forKey: .type)
        try container.encode(payload, forKey: .payload)
        try container.encode(customFields, forKey: .customFields)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
        try container.encode(isDeleted, forKey: .isDeleted)
        try container.encodeIfPresent(deletedAt, forKey: .deletedAt)
        try container.encode(version, forKey: .version)
        try container.encode(updatedBy, forKey: .updatedBy)
    }
}

struct MasterKeyRecord: Codable, Equatable, Sendable {
    var saltBase64: String
    var iterations: Int
    var verifierBase64: String
    var metadataSaltBase64: String?
    var metadataIterations: Int?
}

struct EncryptedPayloadRecord: Codable, Equatable, Sendable {
    var ciphertext: String
    var nonce: String
    var mac: String
    var version: Int
}

struct VaultPersistenceEnvelope: Codable, Equatable, Sendable {
    var schemaVersion: Int = 1
    var masterKeyRecord: MasterKeyRecord?
    var encryptedVault: EncryptedPayloadRecord?
    var updatedAt: Date = Date()
}

struct VaultSnapshot: Codable, Equatable, Sendable {
    var entries: [VaultEntry] = []
    var categories: [String] = []
    var tags: [String] = []
    var security: SecuritySettings = SecuritySettings()
    var syncStatus: String = "Not configured"
    var lastBackupStatus: String = "No backup has run"
    var updatedAt: Date = Date()
}

struct SecuritySettings: Codable, Equatable, Sendable {
    var requireTotp = false
    var totpSecret = ""
}

enum ScopedExportScope: String, Codable, Sendable {
    case item
    case category
}

struct ScopedVaultExport: Codable, Equatable, Sendable {
    var version = 1
    var scope: ScopedExportScope
    var exportedAt: Date
    var item: VaultEntry?
    var category: String?
    var items: [VaultEntry]?
}

struct EntryExportField: Identifiable, Hashable, Sendable {
    var id: String
    var title: String
}

enum ImportConflictStrategy: String, CaseIterable, Identifiable, Sendable {
    case keepCopy
    case overwrite
    case skip

    var id: String { rawValue }

    var title: String {
        switch self {
        case .keepCopy: "Keep Copy"
        case .overwrite: "Overwrite"
        case .skip: "Skip"
        }
    }
}

enum VaultFilter: Hashable, Sendable {
    case all
    case type(VaultEntryType)
    case category(String)
    case tag(String)

    var title: String {
        switch self {
        case .all: "All Items"
        case .type(let type): type.title
        case .category(let category): category
        case .tag(let tag): "#\(tag)"
        }
    }
}

extension String {
    var safeExportName: String {
        let invalid = CharacterSet(charactersIn: "\\/:*?\"<>|")
            .union(.whitespacesAndNewlines)
        let parts = components(separatedBy: invalid).filter { !$0.isEmpty }
        return parts.joined(separator: "_").isEmpty ? "untitled" : parts.joined(separator: "_")
    }
}

extension VaultEntry {
    var safeExportName: String {
        label.safeExportName
    }

    var importMatchKey: String {
        "\(type.rawValue)|\(label.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())|\(payload.category.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())"
    }

    func copyForImport(id: UUID, updatedAt: Date) -> VaultEntry {
        VaultEntry(
            id: id,
            label: label,
            type: type,
            payload: payload,
            customFields: customFields,
            createdAt: createdAt,
            updatedAt: updatedAt,
            isDeleted: false,
            deletedAt: nil,
            version: version,
            updatedBy: updatedBy
        )
    }

    var exportFields: [EntryExportField] {
        var fields = [
            EntryExportField(id: "label", title: "Label"),
            EntryExportField(id: "category", title: "Category"),
            EntryExportField(id: "tags", title: "Tags")
        ]

        switch payload {
        case .credential:
            fields += [
                EntryExportField(id: "credential.username", title: "Username"),
                EntryExportField(id: "credential.password", title: "Password"),
                EntryExportField(id: "credential.token", title: "Token"),
                EntryExportField(id: "credential.appId", title: "App ID"),
                EntryExportField(id: "credential.accessKey", title: "Access Key"),
                EntryExportField(id: "credential.secretKey", title: "Secret Key"),
                EntryExportField(id: "credential.notes", title: "Notes")
            ]
        case .server:
            fields += [
                EntryExportField(id: "server.name", title: "Name"),
                EntryExportField(id: "server.ipAddress", title: "IP Address"),
                EntryExportField(id: "server.port", title: "Port"),
                EntryExportField(id: "server.username", title: "Username"),
                EntryExportField(id: "server.password", title: "Password"),
                EntryExportField(id: "server.basicConfig", title: "Config"),
                EntryExportField(id: "server.operatingSystem", title: "OS"),
                EntryExportField(id: "server.location", title: "Location"),
                EntryExportField(id: "server.notes", title: "Notes")
            ]
        case .service:
            fields += [
                EntryExportField(id: "service.name", title: "Name"),
                EntryExportField(id: "service.connectionAddress", title: "Address"),
                EntryExportField(id: "service.connectionPort", title: "Port"),
                EntryExportField(id: "service.accountId", title: "Account ID"),
                EntryExportField(id: "service.serverIds", title: "Servers"),
                EntryExportField(id: "service.accounts", title: "Accounts"),
                EntryExportField(id: "service.notes", title: "Notes")
            ]
        }

        fields += customFields.map {
            EntryExportField(
                id: "custom.\($0.id.uuidString)",
                title: $0.name.isEmpty ? "Custom Field" : $0.name
            )
        }
        return fields
    }

    func keepingExportFields(_ selectedFieldIDs: Set<String>) -> VaultEntry {
        let selectedCustomIDs = Set(
            selectedFieldIDs.compactMap { id -> UUID? in
                guard id.hasPrefix("custom.") else { return nil }
                return UUID(uuidString: String(id.dropFirst("custom.".count)))
            }
        )
        return VaultEntry(
            id: id,
            label: selectedFieldIDs.contains("label") ? label : "",
            type: type,
            payload: payload.keepingExportFields(selectedFieldIDs),
            customFields: customFields.filter { selectedCustomIDs.contains($0.id) },
            createdAt: createdAt,
            updatedAt: updatedAt,
            isDeleted: isDeleted,
            deletedAt: deletedAt,
            version: version,
            updatedBy: updatedBy
        )
    }
}

private extension VaultPayload {
    func keepingExportFields(_ selectedFieldIDs: Set<String>) -> VaultPayload {
        switch self {
        case .credential(var credential):
            if !selectedFieldIDs.contains("category") { credential.category = "" }
            if !selectedFieldIDs.contains("tags") { credential.tags = [] }
            if !selectedFieldIDs.contains("credential.username") { credential.username = "" }
            if !selectedFieldIDs.contains("credential.password") { credential.password = "" }
            if !selectedFieldIDs.contains("credential.token") { credential.token = "" }
            if !selectedFieldIDs.contains("credential.appId") { credential.appId = "" }
            if !selectedFieldIDs.contains("credential.accessKey") { credential.accessKey = "" }
            if !selectedFieldIDs.contains("credential.secretKey") { credential.secretKey = "" }
            if !selectedFieldIDs.contains("credential.notes") { credential.notes = "" }
            return .credential(credential)
        case .server(var server):
            if !selectedFieldIDs.contains("category") { server.category = "" }
            if !selectedFieldIDs.contains("tags") { server.tags = [] }
            if !selectedFieldIDs.contains("server.name") { server.name = "" }
            if !selectedFieldIDs.contains("server.ipAddress") { server.ipAddress = "" }
            if !selectedFieldIDs.contains("server.port") { server.port = "" }
            if !selectedFieldIDs.contains("server.username") { server.username = "" }
            if !selectedFieldIDs.contains("server.password") { server.password = "" }
            if !selectedFieldIDs.contains("server.basicConfig") { server.basicConfig = "" }
            if !selectedFieldIDs.contains("server.operatingSystem") { server.operatingSystem = "" }
            if !selectedFieldIDs.contains("server.location") { server.location = "" }
            if !selectedFieldIDs.contains("server.notes") { server.notes = "" }
            return .server(server)
        case .service(var service):
            if !selectedFieldIDs.contains("category") { service.category = "" }
            if !selectedFieldIDs.contains("tags") { service.tags = [] }
            if !selectedFieldIDs.contains("service.name") { service.name = "" }
            if !selectedFieldIDs.contains("service.connectionAddress") { service.connectionAddress = "" }
            if !selectedFieldIDs.contains("service.connectionPort") { service.connectionPort = "" }
            if !selectedFieldIDs.contains("service.accountId") { service.accountId = nil }
            if !selectedFieldIDs.contains("service.serverIds") { service.serverIds = [] }
            if !selectedFieldIDs.contains("service.accounts") { service.accounts = [] }
            if !selectedFieldIDs.contains("service.notes") { service.notes = "" }
            return .service(service)
        }
    }
}

File: apps/android_native/app/src/main/java/life/devops/passwordmanager/model/VaultModels.kt
package life.devops.passwordmanager.model

import java.time.Instant
import java.util.UUID

enum class VaultEntryType(val title: String) {
    CREDENTIAL("Credential"),
    SERVER("Server"),
    SERVICE("Service")
}

data class CredentialPayload(
    val username: String = "",
    val password: String = "",
    val accounts: List<ServiceAccount> = emptyList(),
    val token: String = "",
    val appId: String = "",
    val accessKey: String = "",
    val secretKey: String = "",
    val notes: String = "",
    val tags: List<String> = emptyList(),
    val category: String = "",
)

data class ServerPayload(
    val name: String = "",
    val ipAddress: String = "",
    val port: String = "",
    val username: String = "",
    val password: String = "",
    val accounts: List<ServiceAccount> = emptyList(),
    val basicConfig: String = "",
    val operatingSystem: String = "",
    val location: String = "",
    val notes: String = "",
    val tags: List<String> = emptyList(),
    val accountId: String? = null,
    val category: String = "",
)

data class SecuritySettings(
    val requireTotp: Boolean = false,
    val totpSecret: String = "",
)

data class CategorySyncState(
    val name: String,
    val isDeleted: Boolean = false,
    val updatedAt: Instant = Instant.now(),
    val version: Map<String, Int> = emptyMap(),
    val updatedBy: String = "",
)

data class VaultSnapshot(
    val entries: List<VaultEntry> = emptyList(),
    val categories: List<String> = emptyList(),
    val categoryTemplates: List<CategoryTemplate> = emptyList(),
    val categoryStates: List<CategorySyncState> = emptyList(),
    val tags: List<String> = emptyList(),
    val security: SecuritySettings = SecuritySettings(),
    val syncStatus: String = "Not configured",
    val backupStatus: String = "No backup has run",
    val updatedAt: Instant = Instant.now(),
)

data class ServiceAccount(
    val username: String = "",
    val password: String = "",
    val note: String = "",
)

data class ServicePayload(
    val name: String = "",
    val connectionAddress: String = "",
    val connectionPort: String = "",
    val accountId: String? = null,
    val serverIds: List<String> = emptyList(),
    val accounts: List<ServiceAccount> = emptyList(),
    val notes: String = "",
    val tags: List<String> = emptyList(),
    val category: String = "",
)

data class CustomField(
    val id: String = UUID.randomUUID().toString(),
    val name: String = "",
    val value: String = "",
    val templateFieldId: String = "",
)

data class FieldTemplate(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val valueType: String = "text",
    val targetCategory: String = "",
    val targetFieldId: String = "",
)

data class CategoryTemplate(
    val category: String,
    val fields: List<FieldTemplate> = defaultCategoryFields(),
) {
    companion object {
        fun defaultCategoryFields(): List<FieldTemplate> =
            listOf(
                FieldTemplate(stableFieldId("名称"), "名称"),
                FieldTemplate(stableFieldId("备注"), "备注"),
            )

        fun fieldsForPreset(
            preset: CategoryTypePreset?,
            customFieldNames: List<String> = emptyList(),
        ): List<FieldTemplate> =
            (defaultCategoryFields() + preset.orEmptyFields().map { fieldName ->
                FieldTemplate(stableFieldId(fieldName), fieldName)
            } + customFieldNames.map { fieldName ->
                FieldTemplate(stableFieldId(fieldName), fieldName)
            })
                .filter { it.name.trim().isNotEmpty() }
                .distinctBy { it.name.trim().lowercase() }

        internal fun stableFieldId(name: String): String {
            val trimmed = name.trim()
            val normalized = trimmed.lowercase()
                .replace(Regex("""[^a-z0-9\u4e00-\u9fff]+"""), "_")
                .trim('_')
            val suffix = when {
                normalized.isNotEmpty() -> normalized
                trimmed.isEmpty() -> "empty"
                else -> "u_${trimmed.toByteArray(Charsets.UTF_8).joinToString("") { byte ->
                    (byte.toInt() and 0xff).toString(16).padStart(2, '0')
                }}"
            }
            return "template_$suffix"
        }

        private fun CategoryTypePreset?.orEmptyFields(): List<String> =
            this?.fields ?: emptyList()
    }
}

enum class CategoryTypePreset(val title: String, val fields: List<String>) {
    SERVER("服务器", listOf("IP地址", "端口", "关联账号")),
    SERVICE("服务", listOf("服务入口", "关联账号", "关联服务器")),
    ACCOUNT("账号", listOf("入口"));

    companion object {
        fun fromTitle(value: String): CategoryTypePreset? =
            entries.firstOrNull { it.title == value.trim() }
    }
}

sealed interface VaultPayload {
    val category: String
    val tags: List<String>

    data class Credential(val value: CredentialPayload) : VaultPayload {
        override val category: String = value.category
        override val tags: List<String> = value.tags
    }

    data class Server(val value: ServerPayload) : VaultPayload {
        override val category: String = value.category
        override val tags: List<String> = value.tags
    }

    data class Service(val value: ServicePayload) : VaultPayload {
        override val category: String = value.category
        override val tags: List<String> = value.tags
    }
}

data class VaultEntry(
    val id: String = UUID.randomUUID().toString(),
    val label: String,
    val type: VaultEntryType,
    val payload: VaultPayload,
    val customFields: List<CustomField> = emptyList(),
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
    val version: Map<String, Int> = emptyMap(),
    val updatedBy: String = "android",
    val isDeleted: Boolean = false,
    val deletedAt: Instant? = null,
)

val String.safeExportName: String
    get() {
        val invalid = Regex("""[\\/:*?"<>|\s]+""")
        val safeName = trim()
            .split(invalid)
            .filter { it.isNotEmpty() }
            .joinToString("_")
        return safeName.ifEmpty { "untitled" }
    }

val VaultEntry.safeExportName: String
    get() = label.safeExportName

val VaultEntry.importMatchKey: String
    get() = listOf(
        type.name.lowercase(),
        label.trim().lowercase(),
        payload.category.trim().lowercase(),
    ).joinToString("|")

fun VaultEntry.copyForImport(id: String = UUID.randomUUID().toString(), updatedAt: Instant = Instant.now()): VaultEntry =
    copy(
        id = id,
        updatedAt = updatedAt,
        isDeleted = false,
        deletedAt = null,
    )

data class MasterKeyRecord(
    val saltBase64: String,
    val iterations: Int,
    val verifierBase64: String,
    val metadataSaltBase64: String? = null,
    val metadataIterations: Int? = null,
)

data class EncryptedPayloadRecord(
    val ciphertext: String,
    val nonce: String,
    val mac: String,
    val version: Int,
)

data class VaultPersistenceEnvelope(
    val schemaVersion: Int = 1,
    val masterKeyRecord: MasterKeyRecord? = null,
    val encryptedVault: EncryptedPayloadRecord? = null,
    val updatedAt: Instant = Instant.now(),
)

enum class ScopedExportScope {
    ITEM,
    CATEGORY
}

data class ScopedVaultExport(
    val version: Int = 2,
    val scope: ScopedExportScope,
    val exportedAt: Instant = Instant.now(),
    val item: VaultEntry? = null,
    val category: String? = null,
    val items: List<VaultEntry>? = null,
    val categoryTemplates: List<CategoryTemplate> = emptyList(),
)

enum class ImportConflictStrategy(val title: String) {
    KEEP_COPY("Keep Copy"),
    OVERWRITE("Overwrite"),
    SKIP("Skip")
}

data class EntryDraft(
    val label: String,
    val type: VaultEntryType,
    val category: String,
    val tags: List<String>,
    val customFields: List<CustomField> = emptyList(),
    val credential: CredentialPayload = CredentialPayload(),
    val server: ServerPayload = ServerPayload(),
    val service: ServicePayload = ServicePayload(),
) {
    fun toPayload(): VaultPayload = when (type) {
        VaultEntryType.CREDENTIAL -> VaultPayload.Credential(
            credential.copy(category = category, tags = tags)
        )
        VaultEntryType.SERVER -> VaultPayload.Server(
            server.copy(category = category, tags = tags)
        )
        VaultEntryType.SERVICE -> VaultPayload.Service(
            service.copy(category = category, tags = tags)
        )
    }
}
